src/gromacs/nblib/bondtypes.cpp

```diff
@@ -44,25 +44,120 @@
  */
 #include "gmxpre.h"
 
+#include "util.h"
 #include "bondtypes.h"
 
 namespace nblib
 {
 
+namespace detail
+{
+
+template<class B, class... Ps>
+bool commonCompare(const B& a, const B& b, const std::tuple<Ps...>& properties)
+{
+    auto elementComparison = transform_tuple([&a, &b](const auto& f) { return f(a) == f(b); }, properties);
+    auto logicalAnd        = [](auto&& a, auto&& b) { return a && b; };
+    return std17::apply([f = logicalAnd](auto&&... args) { return binary_fold(f, args...); },
+                        elementComparison);
+}
+
+template<class B, class Op, class... Ps>
+bool commonRelational(const B& a, const B& b, Op&& op, const std::tuple<Ps...>& properties)
+{
+    auto compare = [](const auto& a, const auto& b, auto&& op, const auto&... properties) {
+        auto start = [](auto&& recurse, const auto& a, const auto& b, auto&& op,
+                        const auto& property, const auto&... properties) {
+            if (op(property(a), property(b)))
+            {
+                return true;
+            }
+            else if (op(property(a), property(b)))
+            {
+                return false;
+            }
+            else
+            {
+                return std::get<(sizeof...(properties) == 0)>(recurse)(
+                        std::forward<decltype(recurse)>(recurse), a, b,
+                        std::forward<decltype(op)>(op), properties...);
+            }
+        };
+
+        auto stop = [](auto&&... args) {
+            ignore_unused(args...);
+            return false;
+        };
+
+        return start(std::make_tuple(start, stop), a, b, std::forward<decltype(op)>(op), properties...);
+    };
+
+    return std17::apply(
+            [&a, &b, &op, f = compare](const auto&... args) {
+                return f(a, b, std::forward<Op>(op), args...);
+            },
+            properties);
+}
+
+} // namespace detail
+
 HarmonicBondType::HarmonicBondType(Name name, ForceConstant forceConstant, EquilDistance equilDistance) :
     name_(name),
     forceConstant_(forceConstant),
     equilDistance_(equilDistance)
 {
 }
 
+auto HarmonicBondType::getProperties()
```

Comment:
The power of the "crazy" variadic templated lambda code above is easily displayed, because the only distinction between the `less than` and `equal` operators is the call to the specific `getProperties` functions for the `HarmonicBondType `, `G96BondType`, `CubicBondType`, `FENEBondType`, `MorseBondType` and `HalfAttractiveQuarticBondType`.
Loved it!